Xamarin.Forms.Core.UnitTests/DataTemplateSelectorTests.cs

```diff
@@ -90,4 +90,69 @@ public void NestingThrowsException ()
 			Assert.Throws<NotSupportedException> (() => dts.SelectTemplate ((byte)0, null));
 		}
 	}
+
+	[TestFixture]
+	public class DataTemplateRecycleTests : BaseTestFixture
+	{
+		[TearDown]
+		public override void TearDown()
+		{
+			base.TearDown();
+			Device.PlatformServices = null;
+		}
+
+		[SetUp]
+		public override void Setup()
+		{
+			base.Setup();
+			Device.PlatformServices = new MockPlatformServices();
+		}
+
+		class TestDataTemplateSelector : DataTemplateSelector
+		{
+			readonly DataTemplate declarativeTemplate;
+			readonly DataTemplate proceduralTemplate;
+
+			public TestDataTemplateSelector ()
+			{
+				declarativeTemplate = new DataTemplate(typeof(ViewCell));
+				proceduralTemplate = new DataTemplate(() => new EntryCell());
+			}
+
+			protected override DataTemplate OnSelectTemplate(object item, BindableObject container)
+			{
+				Counter++;
+
+				if (item is string)
+					return declarativeTemplate;
```

Comment:
Given that (1) the DataTemplate is purely a function of the type of the item and (2) the DataTemplate declares it always returns a ViewCell (e.g. uses the Type ctor) let us know that all strings will be rendered with a ViewCell and allows us to cache that fact.